### ISSUE TITLE
A question about this work.

### ISSUE THREAD

**Issue by OuyangJunyuan** [lly007/PointStructuringNet#1]
As mentioned in the paper, Gumbel Softmax is used to perform argmax operation in a differentiable way.
Thus the `sub` set obtained by Eq. (11) could be no exist in the origin input points during the training phase, right?

**Comment by lly007** [comment]
Indeed.
During the research of this paper, we have considered using 3 argmax methods in the training phase.
1. directly using argmax, the problem of this method is that it cannot be trained and the weights of SFTF are completely randomized. However, the performance of this method is not bad, close to or slightly better than the original method on PointNet++ and PointConv, but the performance is less stable.
2. With Gumbel Softmax, a smaller temperature value is used throughout the training phase, and the network can still be trained, but the points in sub do not exactly correspond to the original point cloud, but are only very close to some points in the original point cloud, but the convergence time is slightly longer.
3. Using Gumbel Softmax, we set a temperature curve in the training phase and turn it into a hardmax in the later stage of training, which also performs well.
--------------------------------------------------------------
也可以汉语交流。
事实上正是这样。
在本文研究过程中，我们曾经考虑过在训练阶段使用3种argmax方法。
1. 直接使用argmax，这种方法的问题在于无法训练，SFTF的权重完全随机化。然而，这种方法的性能并不差，在PointNet++和PointConv上接近或稍优于原始方法，但是表现不太稳定。
2. 使用Gumbel Softmax，整个训练阶段使用一个较小的temperature值，网络依然可以训练，但是sub中的点不能完全对应原始点云，只是非常接近原始点云中的某些点，但是收敛周期略长。
3. 使用Gumbel Softmax，训练阶段设置一个temperature曲线，在训练的后期变成一个hardmax，表现也还不错。

**Comment by OuyangJunyuan** [comment]
感谢回复，为何说argmax无法训练呢，我看好几多文章都那么说（被选择的那些点应该还是有梯度），基础薄弱请多指教。还有就是运行您的代码发现`pointnet_utils_psn.py-178`这行报错实参过多，这份代码是否不是最终可运行的代码呢。

**Comment by lly007** [comment]
关于第一个问题：
我举个例子
比如说有一个向量 a=[3,4,1,2,7]，我对他进行一个sort，会得到2个向量，一个是sort的结果[7,4,3,2,1]，一个是sort的index [4,1,0,3,2]
sort的结果你可以认为是有梯度的，但是我们并不关心这个结果，因为这个结果实际上是由index发现的，重点问题还是在于index。
这个index显然是不可导的。更直观一点解释，假设我现在需要最大的index不是4，而是1，也就是a中的4，这时反向传播会对a产生一个误差，也就是让a中7变小，让4变大，但是sort产生的index无法具体给出如何让7变小，它只是一个index用来指示谁更大而已。
因此，本文用了gumbel。gumbel的原理不再详述，我只说一下直观的效果。
比如说我们想用gumbel得到最大值，直接对a求gumbel，可能会得到一个类似于g=[0.001,0.001,0.0005,0.0008,0.9967]的向量。直接用这个向量去和a求内积会得到一个非常接近于7的数字。如果不希望索引为4的位置最大，反向传播的时候，g的元素产生误差，就可以修改最大值的索引。这就是用gumbel来替代argmax的原理。
也可以直接使用gumbel hardmax，这是一个小trick，效果类似。

第二个问题：我记得是可以运行的，当然也可能submit的时候有一些小bug没处理，你先尝试解决一下，最近在写学位论文有点忙，等闲下来一定处理好。

你也可以直接加我QQ方便交流，276029693，相互学习。

**Comment by lly007** [comment]
The bug has been fixed.
